yt_dlp/utils/traversal.py

```diff
@@ -277,13 +299,115 @@ def _traverse_obj(obj, path, allow_empty, test_type):
         return results[0] if results else {} if allow_empty and is_dict else None
 
     for index, path in enumerate(paths, 1):
-        result = _traverse_obj(obj, path, index == len(paths), True)
-        if result is not None:
-            return result
+        is_last = index == len(paths)
+        try:
+            result = _traverse_obj(obj, path, is_last, True)
+            if result is not None:
+                return result
+        except _RequiredError as e:
+            if is_last:
+                # Reraise to get cleaner stack trace
+                raise ExtractorError(e.orig_msg, expected=e.expected) from None
 
     return None if default is NO_DEFAULT else default
 
 
+def require(name, /, *, expected=False):
+    def func(value):
+        if value is None:
+            raise _RequiredError(f'Unable to extract {name}', expected=expected)
+
+        return value
+
+    return func
+
+
+class _RequiredError(ExtractorError):
+    pass
+
+
+@typing.overload
+def subs_list_to_dict(*, ext: str | None = None) -> collections.abc.Callable[[list[dict]], dict[str, list[dict]]]: ...
+
+
+@typing.overload
+def subs_list_to_dict(subs: list[dict] | None, /, *, ext: str | None = None) -> dict[str, list[dict]]: ...
+
+
+def subs_list_to_dict(subs: list[dict] | None = None, /, *, ext=None):
+    """
+    Convert subtitles from a traversal into a subtitle dict.
+    The path should have an `all` immediately before this function.
+
+    Arguments:
+    `ext`      The default value for `ext` in the subtitle dict
+
+    In the dict you can set the following additional items:
+    `id`       The subtitle id to sort the dict into
+    `quality`  The sort order for each subtitle
+    """
+    if subs is None:
+        return functools.partial(subs_list_to_dict, ext=ext)
+
+    result = collections.defaultdict(list)
+
+    for sub in subs:
+        if not url_or_none(sub.get('url')) and not sub.get('data'):
+            continue
+        sub_id = sub.pop('id', None)
+        if sub_id is None:
+            continue
+        if ext is not None and not sub.get('ext'):
+            sub['ext'] = ext
+        result[sub_id].append(sub)
+    result = dict(result)
+
+    for subs in result.values():
+        subs.sort(key=lambda x: x.pop('quality', 0) or 0)
+
+    return result
+
+
+def find_element(tag=None, id=None, cls=None, attr=None, value=None, html=False):
+    # deliberately using `id=` and `cls=` for ease of readability
```

Comment:
Undeniably readable but I think I'm missing the point that this and the similar comment below are making. As opposed to what? `_id=`? Positional-only params?

Or should some/all of the params be kw-only?

Reply:
`id_`, `cls_`. Our linter is configured against these names, so when we eventually lint core code this should be kept, and exceptions added. Kw only makes sense though

Reply:
OK, I've now had a reasonably successful go at back-porting this (filter_element[s]). I think it would be better to fix/improve some of the underlying utils code before creating tests and committing this. Maybe omit them from this PR and instead have 1 or 2 PRs that first fix/improve the get_element...(), etc functions and then re-implement filter_element[s]?

utils changes that I would propose upstream:
* Fix the _by_class_ fns so that class matching works like CSS: `foo bar` should match `class="foo bar"` but also `class="bar foo"`, `class="foo baz bar", etc. My implementation for this works by allowing `value` also to be a `(str) --> bool` filter.
* Allow matching by `tag` everywhere using kwarg `tag=`. Just by adding `**kwargs` to all the top-level fns we can also support `escape_tag=` like `escape_value`, though it's less relevant since tag-names won't normally contain RE special characters (`.` ?).
* Allow matching empty value as in `<div ... itemscope>` or `<div ... itemscope="">`.
* Allow matching self-closing tags in 2 ways -- currently this works if such a tag is embedded like `<div ...><img ...></div>` but fails with a fragment containing a top-level self-closing tag, as might be generated by nested calls:
  1. don't expect to close a `<tag ... />` (believing the markup even if it isn't not really valid HTML): fixed by adding `startendtag()` method to `HTMLBreakOnClosingTagParser` that raises `HTMLBreakOnClosingTagException`.
  2. don't expect to close a void tag: fixed by redirecting `starttag()` to above if the tag is in the known list of void tags, and discarding any end tag `</img>` forpe a void tag.
* Py3's `html` module has HTML5 implementations of `escape()` and `unescape()` that should be called instead of the existing code in `escapeHTML()` and `unescapeHTML()` (I have a compat for Py2), so addressing `&eacute`, etc; but a numeric entity with invalidly large codepoint must be replaced with `'\uFFFD'` rather than left as-is as our code does now.

From looking at the `html.parser.HTMLParser` CPython code, it doesn't appear to know about the [HTML5 rules for omitting certain start and end tags under certain conditions](https://html.spec.whatwg.org/multipage/syntax.html#optional-tags). Although it would be good to teach `HTMLBreakOnClosingTagParser` these rules, I can't see how to do that within the `HTMLParser` base class framework: the rules require visibility of, eg, parent, child and sibling elements as well as raw whitespace, whereas we're only told the tag name.

Reply:
That should be done in a separate PR, since I'd want to merge flashdaggers parser improvements before changing anything to the html parser code utils. His implementation can be used much more easily to feed the backend for the util functions and the new helpers. After this is merged, I'll create a new branch from his and port it to the current structure.

Reply:
I'd forgotten #5588 (it has been pending for a while). I'll prepare an upstream version and see how that works.